File: sdk/src/main/java/com/libreshockwave/cast/ShapeInfo.java
package com.libreshockwave.cast;

import com.libreshockwave.io.BinaryReader;

import java.nio.ByteOrder;

/**
 * Shape-specific cast member data.
 */
public record ShapeInfo(
    ShapeType shapeType,
    int regX,
    int regY,
    int width,
    int height,
    int color,
    int backColor,
    int fillType,
    int lineThickness,
    int lineDirection
) implements Dimensioned {

    public enum ShapeType {
        RECT(0x01),
        OVAL_RECT(0x02),
        OVAL(0x03),
        LINE(0x08),
        UNKNOWN(0);

        private final int code;

        ShapeType(int code) {
            this.code = code;
        }

        public static ShapeType fromCode(int code) {
            for (ShapeType type : values()) {
                if (type.code == code) return type;
            }
            return UNKNOWN;
        }
    }

    public static ShapeInfo parse(byte[] data) {
        if (data == null || data.length < 14) {
            return new ShapeInfo(ShapeType.UNKNOWN, 0, 0, 0, 0, 0, 0, 0, 1, 0);
        }

        BinaryReader reader = new BinaryReader(data, ByteOrder.BIG_ENDIAN);

        int shapeTypeRaw = reader.readU16();
        int regY = reader.readU16();
        int regX = reader.readU16();
        int height = reader.readU16();
        int width = reader.readU16();
        reader.skip(2); // unknown
        int color = reader.readU8();
        int backColor = data.length >= 14 ? data[13] & 0xFF : 0;
        int fillType = data.length >= 15 ? data[14] & 0xFF : 1;
        int lineThickness = data.length >= 16 ? data[15] & 0xFF : 1;
        int lineDirection = data.length >= 17 ? data[16] & 0xFF : 0;

        return new ShapeInfo(
            ShapeType.fromCode(shapeTypeRaw),
            regX, regY,
            width, height,
            color,
            backColor,
            fillType,
            lineThickness,
            lineDirection
        );
    }

    public boolean isFilled() {
        return fillType != 0;
    }

    public boolean isOutlineInvisible() {
        return !isFilled() && lineThickness <= 1;
    }
}

File: cpp/src/cast/ShapeInfo.cpp
#include "libreshockwave/cast/ShapeInfo.hpp"

#include "libreshockwave/io/BinaryReader.hpp"

namespace libreshockwave::cast {

ShapeType shapeTypeFromCode(int value) {
    switch (value) {
        case 0x01: return ShapeType::Rect;
        case 0x02: return ShapeType::OvalRect;
        case 0x03: return ShapeType::Oval;
        case 0x08: return ShapeType::Line;
        default: return ShapeType::Unknown;
    }
}

bool ShapeInfo::isFilled() const {
    return fillType != 0;
}

bool ShapeInfo::isOutlineInvisible() const {
    return !isFilled() && lineThickness <= 1;
}

ShapeInfo ShapeInfo::parse(const std::vector<std::uint8_t>& data) {
    if (data.size() < 14) {
        return ShapeInfo{ShapeType::Unknown, 0, 0, 0, 0, 0, 0, 0, 1, 0};
    }

    io::BinaryReader reader(data, io::ByteOrder::BigEndian);
    const int shapeTypeRaw = reader.readU16();
    const int regY = reader.readU16();
    const int regX = reader.readU16();
    const int height = reader.readU16();
    const int width = reader.readU16();
    reader.skip(2);
    const int color = reader.readU8();
    const int backColor = data.size() >= 14 ? data[13] : 0;
    const int fillType = data.size() >= 15 ? data[14] : 1;
    const int lineThickness = data.size() >= 16 ? data[15] : 1;
    const int lineDirection = data.size() >= 17 ? data[16] : 0;

    return ShapeInfo{shapeTypeFromCode(shapeTypeRaw), regX, regY, width, height, color, backColor, fillType, lineThickness, lineDirection};
}

} // namespace libreshockwave::cast
